### PR TITLE
NXP backend: Add prelu support using new neutron flow

### PR DESCRIPTION
**Pull request by roman-janik-nxp**
### Summary
Add tests verifying correct support for prelu by the Neutron backend using the new Neutron MLIR flow

### Test plan
Unit tests provided.

cc @robert-kalmar @JakeStevens @digantdesai @rascani

### CODE REVIEW

**Comment by roman-janik-nxp** [comment]
@novak-vaclav

File: backends/nxp/tests/ir/converter/node_converter/test_add_tensor_converter.py [hunk 1]
```diff
@@ -104,6 +104,10 @@ def test__basic_nsys_inference_qat(self, mocker, request):
             pytest.param(
                 [ModelInputSpec((4,)), ModelInputSpec((4, 4))], id="2 inputs 1D + 2D."
             ),
+            pytest.param(
+                [ModelInputSpec((10,)), ModelInputSpec((1, 1))],
+                id="2 inputs 2D, num_elems of input == num_elems of output",
+            ),
```
[the remark this answers is not in the document]
Comment (novak-vaclav) [reply backends/nxp/tests/ir/converter/node_converter/test_add_tensor_converter.py:121]: nit: true, please fix

File: backends/nxp/tests/ir/converter/node_converter/test_mul_tensor_converter.py [hunk 2]
```diff
@@ -90,6 +90,10 @@ def test__basic_nsys_inference_qat(self, mocker, request, x_input_shape):
             pytest.param(
                 [ModelInputSpec((4,)), ModelInputSpec((4, 4))], id="2 inputs 1D+2D."
             ),
+            pytest.param(
+                [ModelInputSpec((10,)), ModelInputSpec((1, 1))],
+                id="2 inputs 2D, num_elems of input == num_elems of output",
+            ),
```
[the remark this answers is not in the document]
Comment (novak-vaclav) [reply backends/nxp/tests/ir/converter/node_converter/test_mul_tensor_converter.py:102]: nit: true, please fix

File: backends/nxp/tests/ir/converter/node_converter/test_sub_tensor_converter.py [hunk 3]
```diff
@@ -104,6 +104,10 @@ def test__basic_nsys_inference_qat(self, mocker, request):
                 [ModelInputSpec((5, 3, 4)), ModelInputSpec((1, 3, 1))],
                 id="2 inputs 3D.",
             ),
+            pytest.param(
+                [ModelInputSpec((10,)), ModelInputSpec((1, 1))],
+                id="2 inputs 2D, num_elems of input == num_elems of output",
+            ),
```
[the remark this answers is not in the document]
Comment (novak-vaclav) [reply backends/nxp/tests/ir/converter/node_converter/test_sub_tensor_converter.py:117]: Nit: true, please fix

File: backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py [hunk 4]
```diff
@@ -29,123 +43,181 @@ def reseed_model_per_test_run():
...
-        ],
-    )
-
-    assert graph_contains_any_of_ops(
-        edge_program.graph,
+    def test__basic_nsys_inference(self, mocker, request, input_shape):
+        channels = input_shape[-1]
+        rank = len(input_shape)
+        model = LinearPReLUModule(in_features=channels, out_features=channels)
+        graph_verifier = DetailedGraphVerifier(
```
Comment (novak-vaclav) [inline backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py]: The code of the test cases seem to have a lot in common, I'd try to extract some of the functionality into a separate method

**Review by novak-vaclav — COMMENTED** [review]
A few minor comments, otherwise LGTM. Good job 😄

File: backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py (hunk 4 above)
Comment (roman-janik-nxp) [reply backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py]: Done.

**Review by novak-vaclav — APPROVED** [review]
LGTM 👌

File: backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py [hunk 5]
```diff
@@ -29,123 +43,148 @@ def reseed_model_per_test_run():
...
-    ],
-)
-def test_prelu_with_linear_quant_conversion(mocker, input_shape):
-    converter_spy = mocker.spy(EdgeProgramToIRConverter, "convert_program")
-
-    # Run conversion
-    channels = input_shape[-1]
-    edge_program = to_quantized_edge_program(
-        LinearPReLUModule(in_features=channels, out_features=channels),
+class TestPreluConverter:
```
Comment (MartinPavella) [inline backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py:50]: The PyTorch documentation states that the `num_parameters` parameter can either be `1`, or it can be equal to the number of **channels**. Since the **channels** are important, we must identify how this operator behaves with respect to the NCHW -> NHWC conversion in our pipeline. I don't see any code in the converter addressing node formats, and I don't see any tests here using the channels first format and `num_parameters != 1`. Please add such tests, and make any necessary changes to the `PReluConverter`.

Image

**Comment by roman-janik-nxp** [comment]
@MartinPavella I added conditional Transpose ops for the channels first format and `num_parameters != 1` case. I don't think there should be any changes in `NodeFormatInference` as we discussed today. Please check again.

File: backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py [hunk 6]
```diff
@@ -29,123 +43,156 @@ def reseed_model_per_test_run():
...
+
+        self.assert_delegated(model, input_shape, mocker, request, use_qat=True)
+
+    def test__num_parameters_param(self, mocker, request):
+        input_shape = (1, 43, 93)
+        channels = input_shape[-1]
+        num_parameters = input_shape[1]
+        model = LinearPReLUModule(
+            in_features=channels, out_features=channels, num_parameters=num_parameters
+        )
```
Comment (MartinPavella) [inline backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py]: I am a bit confused by the behavior of the `aten.prelu` and the Neutron IR `PRelu`, could you please clear some things up for me?

In this test, the `aten.prelu` uses `num_parameters=43`, which corresponds to the second dimension of it's input. When this model is converted to Neutron IR, the input shape remains unchanged (as there is no conv/polling/...), so the Neutron IR `PRelu` has `43` parameters, which also corresponds to the second dimension.

Then in the `test__channels_first` test, there is a `conv` in the model, so the input shape is channels first and the `aten.prelu` has 8 parameters, which corresponds to the second dimension of its input (the channels). In Neutron IR, the input shape will be permuted to channels last ([1,42, 24, 8] to be exact), and the parameters would correspond to the last dimension (channels in NHWC), which (I assume from the implementation) would not work. It appears that the Neutron IR `Prelu` requires the parameters to correspond to the second dimension (not the channels), and that's why you have added the code to permute the input to channels first.

Is my understanding correct?

Comment (roman-janik-nxp) [reply backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py]: I'm sorry, I made a mistake somewhere. The code for adding Transposes is actually not needed at all. The PyTorch requires the `num_parameters` to be `[channels]` or `[1]`. The Neutron IR requires the tensor created from `num_parameters` to be same size as input (not applicable) or broadcastable. `[1]` is broadcastable trivially, and the `[channels]` has here is the last dim [x, x, x, channels] * [channels] - the broadcast is done from right.

For formatless input it's a little bit more complicated. The shape is [1, 43, 93] = [N, C, L], so the channels for Prelu is 43 (even without Linear op). This makes it [1, 43, 93] * [43] which is not broadcastable in Neutron IR execution (checked it), but it's ok for Neutron converter.

Comment (MartinPavella) [reply backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py]: All right, thanks for the explanation.
In that case, we **need to handle the `prelu` during node format inference as an operator that requires the channels first format** if `num_parameters != 1` (if `num_parameters == 1`, we must handle it as an "operator that can use any format").
Your point that [1, 43, 93] * [43] is not broadcastable but it's OK because Neutron Converter can handle it, is very fragile. For example what happens with [1, 8, 8] * [8]? You need to identify these edge cases. We cannot rely on "strange" behavior of the Neutorn Converter.

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/prelu_converter.py [hunk 7]
```diff
@@ -65,6 +45,11 @@ def _is_supported_in_IR(
         if len(node.args) != 2:
             return False
 
+        if tuple(node.all_input_nodes[1].meta["val"].shape) != (1,):
+            return (
+                node.all_input_nodes[0].meta["nxp_node_format"]
+                == node.meta["nxp_node_format"]
+            )
```
Comment (MartinPavella) [inline backends/nxp/backend/ir/converter/node_converters/ops_converters/prelu_converter.py]: What is the purpose of this check? 
As `prelu` is not in the set of ops that can change node format, the returned value just always true.

I assume you meant:
```python
if tuple(node.all_input_nodes[1].meta["val"].shape) != (1,):
    if not node.meta[NXP_NODE_FORMAT].is_channels_first():
        # Should never happen.
        return False
```

File: backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py [hunk 8]
```diff
@@ -29,123 +47,188 @@ def reseed_model_per_test_run():
...
+        self, mocker, request, input_shape, num_parameters_channels
+    ):
+        channels = input_shape[1]
+        num_parameters = channels if num_parameters_channels else 1
+        model = ConvPReLUModule(in_channels=channels, num_parameters=num_parameters)
+        expected_delegated_ops = {
+            Prelu: 1,
+            Convolution: 1,
+        }
+
```
Comment (MartinPavella) [inline backends/nxp/tests/ir/converter/node_converter/test_prelu_converter.py:235]: Could you just add an assert here to make sure the `model.prelu.weight` indeed contains different values?

File: backends/nxp/tests/models.py [hunk 9]
```diff
@@ -902,12 +902,42 @@ def __init__(self, in_features, out_features, num_parameters=1):
...
+class ConvPReLUModule(torch.nn.Module):
+    def __init__(self, in_channels, num_parameters=1):
+        super().__init__()
+
+        self.conv = Conv2dModule(
+            in_channels=in_channels, out_channels=in_channels, stride=1, padding=1
+        )
+        self.prelu = torch.nn.PReLU(num_parameters)
+        # Initialize the weight alpha with random values to simulate learned model with non-constant data
+        torch.nn.init.uniform(self.prelu.weight, -1.0, 1.0)
```
Comment (MartinPavella) [inline backends/nxp/tests/models.py:934]: Nit: Would be cleaner and less error-prone to use the `PReluModule` here, instead of `torch.nn.PReLU` and calling the `torch.nn.init.uniform`. Same for `LinearPReLUModule`. 

But no change necessary. It's up to you.

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/prelu_converter.py (hunk 7 above)
Comment (roman-janik-nxp) [reply backends/nxp/backend/ir/converter/node_converters/ops_converters/prelu_converter.py]: Yes, that's the correct type of check. Actually it should be:
```
if input_rank(node, 0) > 2 and tuple(node.all_input_nodes[1].meta["val"].shape) != (1,):
            if not (node.meta[NXP_NODE_FORMAT].is_channels_first() or node.meta[NXP_NODE_FORMAT] == DataFormat.NONE):
            # Should never happen.
                return False
```
It is valid for higher tensor ranks. Also `DataFormat.None` needs to be accepted as well, because this function is run before partitioner and Node format inference. 

Actually I think its not necessary and prefer to omit it. What you think?

Comment (MartinPavella) [reply backends/nxp/backend/ir/converter/node_converters/ops_converters/prelu_converter.py]: Have you figured out why the NodeFormatInference assigns the node the `DataFormat.None` when run in the aten dialect (before partitioning)?

Comment (roman-janik-nxp) [reply backends/nxp/backend/ir/converter/node_converters/ops_converters/prelu_converter.py]: Thanks for guide, the Prelu needs to be handled also for aten dialect. I modified the code accordingly.

**Review by MartinPavella — APPROVED** [review]
